File: crypto/src/main/java/io/oversec/one/crypto/sym/ui/SymmetricTextEncryptionInfoFragment.java
package io.oversec.one.crypto.sym.ui;

import android.app.Activity;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.text.format.DateUtils;
import android.view.*;
import android.widget.Button;
import android.widget.TextView;
import io.oversec.one.crypto.*;
import io.oversec.one.crypto.sym.OversecKeystore2;
import io.oversec.one.crypto.sym.SymUtil;
import io.oversec.one.crypto.sym.SymmetricCryptoHandler;
import io.oversec.one.crypto.symbase.SymmetricDecryptResult;
import io.oversec.one.crypto.ui.AbstractTextEncryptionInfoFragment;
import io.oversec.one.crypto.ui.EncryptionInfoActivity;

import java.util.Date;

public class SymmetricTextEncryptionInfoFragment extends AbstractTextEncryptionInfoFragment {

    public static SymmetricTextEncryptionInfoFragment newInstance(String packagename) {
        SymmetricTextEncryptionInfoFragment fragment = new SymmetricTextEncryptionInfoFragment();
        fragment.setArgs(packagename);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        mView = inflater.inflate(R.layout.encryption_info_text_sym, container, false);
        super.onCreateView(inflater, container, savedInstanceState);
        return mView;
    }


    @Override
    public void setData(EncryptionInfoActivity activity, String encodedText, BaseDecryptResult tdr, UserInteractionRequiredException uix, final AbstractCryptoHandler encryptionHandler) {
        super.setData(activity, encodedText, tdr, uix, encryptionHandler);

        SymmetricDecryptResult r = (SymmetricDecryptResult) tdr;

        TextView lblSymKeyAlias = (TextView) mView.findViewById(R.id.lbl_sym_key_name);
        TextView tvAvatar = (TextView) mView.findViewById(R.id.tvAvatar);
        TextView tvSym = (TextView) mView.findViewById(R.id.tv_sym_key_name);
        TextView lblConfirm = (TextView) mView.findViewById(R.id.lbl_key_confirm);
        TextView tvConfirm = (TextView) mView.findViewById(R.id.tv_key_confirm);

        if (tdr == null) {

            lblSymKeyAlias.setVisibility(View.GONE);
            tvSym.setVisibility(View.GONE);
            tvAvatar.setVisibility(View.GONE);

        } else {
            final Long keyId = r.getSymmetricKeyId();

            if (keyId == null) {
                lblSymKeyAlias.setVisibility(View.GONE);
                tvSym.setVisibility(View.GONE);
                tvAvatar.setVisibility(View.GONE);
                lblConfirm.setVisibility(View.GONE);
                tvConfirm.setVisibility(View.GONE);
            } else {
                OversecKeystore2 keystore = OversecKeystore2.getInstance(getActivity());
                String name = keystore.getSymmetricKeyEncrypted(keyId).getName();
                tvSym.setText(name);

                SymUtil.applyAvatar(tvAvatar, name);

                Date confirmedDate = keystore.getConfirmDate(keyId);
                tvConfirm.setText(confirmedDate == null ? getActivity().getString(R.string.label_key_unconfirmed) : DateUtils.formatDateTime(getActivity(), confirmedDate.getTime(), 0));
                tvConfirm.setCompoundDrawablesWithIntrinsicBounds(0, 0, confirmedDate == null ? R.drawable.ic_warning_black_24dp : R.drawable.ic_done_black_24dp, 0);
                if (confirmedDate == null) {
                    tvConfirm.setTextColor(ContextCompat.getColor(getActivity(),R.color.colorWarning));
                }


            }

        }


        Button btKeyDetails = (Button) mView.findViewById(R.id.btnKeyDetailsSym);

        if (r != null) {
            final Long keyId = r.getSymmetricKeyId();
            if (keyId == null) {
                btKeyDetails.setVisibility(View.GONE);
            } else {
                btKeyDetails.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {

                        KeyDetailsActivity.show(getActivity(), keyId);
                    }
                });
            }


        }
    }

    @Override
    public boolean onCreateOptionsMenu(Activity activity, Menu menu) {
        activity.getMenuInflater().inflate(R.menu.sym_menu_encryption_info, menu);
        return true;
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        menu.findItem(R.id.action_share_sym_base64).setVisible(mTdr != null);

    }

    @Override
    public void onOptionsItemSelected(Activity activity, MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_share_sym_base64) {
            share(activity, SymmetricCryptoHandler.getRawMessageJson(CryptoHandlerFacade.getEncodedData(activity, mOrigText)), activity.getString(R.string.action_share_sym_base64));
        } else {
            super.onOptionsItemSelected(activity, item);
        }
    }
}

File: crypto/src/main/java/io/oversec/one/crypto/sym/ui/SymmetricTextEncryptionInfoFragment.kt
package io.oversec.one.crypto.sym.ui

import android.app.Activity
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.text.format.DateUtils
import android.view.*
import android.widget.Button
import android.widget.TextView
import io.oversec.one.crypto.*
import io.oversec.one.crypto.sym.OversecKeystore2
import io.oversec.one.crypto.sym.SymUtil
import io.oversec.one.crypto.symbase.BaseSymmetricCryptoHandler
import io.oversec.one.crypto.symbase.SymmetricDecryptResult
import io.oversec.one.crypto.ui.AbstractTextEncryptionInfoFragment
import io.oversec.one.crypto.ui.EncryptionInfoActivity

class SymmetricTextEncryptionInfoFragment : AbstractTextEncryptionInfoFragment() {

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        mView = inflater.inflate(R.layout.encryption_info_text_sym, container, false)
        super.onCreateView(inflater, container, savedInstanceState)
        return mView
    }


    override fun setData(
        activity: EncryptionInfoActivity,
        encodedText: String,
        tdr: BaseDecryptResult?,
        uix: UserInteractionRequiredException?,
        encryptionHandler: AbstractCryptoHandler
    ) {
        super.setData(activity, encodedText, tdr, uix, encryptionHandler)



        val lblSymKeyAlias = mView.findViewById<View>(R.id.lbl_sym_key_name) as TextView
        val tvAvatar = mView.findViewById<View>(R.id.tvAvatar) as TextView
        val tvSym = mView.findViewById<View>(R.id.tv_sym_key_name) as TextView
        val lblConfirm = mView.findViewById<View>(R.id.lbl_key_confirm) as TextView
        val tvConfirm = mView.findViewById<View>(R.id.tv_key_confirm) as TextView
        val btKeyDetails = mView.findViewById<View>(R.id.btnKeyDetailsSym) as Button

        if (tdr == null) {
            lblSymKeyAlias.visibility = View.GONE
            tvSym.visibility = View.GONE
            tvAvatar.visibility = View.GONE

        } else {
            val r = tdr as SymmetricDecryptResult
            val keyId = r.symmetricKeyId

            if (keyId == null) {
                btKeyDetails.visibility = View.GONE
            } else {
                btKeyDetails.setOnClickListener { KeyDetailsActivity.show(getActivity(), keyId) }
            }

            if (keyId == null) {
                lblSymKeyAlias.visibility = View.GONE
                tvSym.visibility = View.GONE
                tvAvatar.visibility = View.GONE
                lblConfirm.visibility = View.GONE
                tvConfirm.visibility = View.GONE
            } else {
                val keystore = OversecKeystore2.getInstance(getActivity())
                val name = keystore.getSymmetricKeyEncrypted(keyId)?.name
                tvSym.text = name

                SymUtil.applyAvatar(tvAvatar, name!!)

                val confirmedDate = keystore.getConfirmDate(keyId)
                tvConfirm.text =
                        if (confirmedDate == null) getActivity().getString(R.string.label_key_unconfirmed) else DateUtils.formatDateTime(
                            getActivity(),
                            confirmedDate.time,
                            0
                        )
                tvConfirm.setCompoundDrawablesWithIntrinsicBounds(
                    0,
                    0,
                    if (confirmedDate == null) R.drawable.ic_warning_black_24dp else R.drawable.ic_done_black_24dp,
                    0
                )
                if (confirmedDate == null) {
                    tvConfirm.setTextColor(
                        ContextCompat.getColor(
                            getActivity(),
                            R.color.colorWarning
                        )
                    )
                }


            }

        }

    }

    override fun onCreateOptionsMenu(activity: Activity, menu: Menu): Boolean {
        activity.menuInflater.inflate(R.menu.sym_menu_encryption_info, menu)
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        menu.findItem(R.id.action_share_sym_base64).isVisible = mTdr != null
    }

    override fun onOptionsItemSelected(activity: Activity, item: MenuItem) {
        val id = item.itemId
        if (id == R.id.action_share_sym_base64) {
            share(
                activity,
                BaseSymmetricCryptoHandler.getRawMessageJson(
                    CryptoHandlerFacade.getEncodedData(
                        activity,
                        mOrigText
                    )!!
                )!!,
                activity.getString(R.string.action_share_sym_base64)
            )
        } else {
            super.onOptionsItemSelected(activity, item)
        }
    }

    companion object {

        fun newInstance(packagename: String): SymmetricTextEncryptionInfoFragment {
            val fragment = SymmetricTextEncryptionInfoFragment()
            fragment.setArgs(packagename)
            return fragment
        }
    }
}
